In [1]:
import pandas as pd
from random import choice

Считывание датафрейма. Считайте датафрейм и сделайте при считывании все признаки строковыми (hints: используйте dtype).

In [2]:
df = pd.read_csv('data/spaceship-titanic/train.csv', dtype=str)
df.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

  RoomService FoodCourt ShoppingMall     Spa VRDeck               Name  \
0         0.0       0.0          0.0     0.0    0.0    Maham Ofracculy   
1       109.0       9.0         25.0   549.0   44.0       Juanna Vines   
2        43.0    3576.0          0.0  6715.0   49.0      Altark Susent   
3         0.0    1283.0        371.0  3329.0  193.0       Solam Susent   
4       303.0      70.0        151.0   565.0    2.0  Willy Santantines   

  Transported  
0       False  
1        True  
2       False  
3       False  
4        True

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PassengerId   8693 non-null   object
 1   HomePlanet    8492 non-null   object
 2   CryoSleep     8476 non-null   object
 3   Cabin         8494 non-null   object
 4   Destination   8511 non-null   object
 5   Age           8514 non-null   object
 6   VIP           8490 non-null   object
 7   RoomService   8512 non-null   object
 8   FoodCourt     8510 non-null   object
 9   ShoppingMall  8485 non-null   object
 10  Spa           8510 non-null   object
 11  VRDeck        8505 non-null   object
 12  Name          8493 non-null   object
 13  Transported   8693 non-null   object
dtypes: object(14)
memory usage: 950.9+ KB


- Выведите несколько первых, последних, случайных строк (случайные используя встроенный метод и сгенерируйте массив из случайных индексов, по которым будут браться строки)

In [4]:
df.head(2)

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   

  RoomService FoodCourt ShoppingMall    Spa VRDeck             Name  \
0         0.0       0.0          0.0    0.0    0.0  Maham Ofracculy   
1       109.0       9.0         25.0  549.0   44.0     Juanna Vines   

  Transported  
0       False  
1        True

In [5]:
df.tail(3)

PassengerId HomePlanet CryoSleep     Cabin  Destination   Age    VIP  \
8690     9279_01      Earth     False  G/1500/S  TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False   E/608/S  55 Cancri e  32.0  False   
8692     9280_02     Europa     False   E/608/S  TRAPPIST-1e  44.0  False   

     RoomService FoodCourt ShoppingMall    Spa  VRDeck              Name  \
8690         0.0       0.0       1872.0    1.0     0.0      Fayey Connon   
8691         0.0    1049.0          0.0  353.0  3235.0  Celeon Hontichre   
8692       126.0    4688.0          0.0    0.0    12.0  Propsh Hontichre   

     Transported  
8690        True  
8691       False  
8692        True

In [6]:
df_sample = df.sample(10)
df_sample

PassengerId HomePlanet CryoSleep     Cabin    Destination   Age    VIP  \
3773     4030_05      Earth     False   G/661/S    TRAPPIST-1e   8.0  False   
7018     7464_01      Earth      True  G/1207/P    TRAPPIST-1e  11.0  False   
6898     7306_01      Earth     False   E/471/P    55 Cancri e  26.0  False   
4157     4441_03     Europa     False   B/144/P    55 Cancri e  50.0  False   
1843     1971_01      Earth     False   F/375/S  PSO J318.5-22  54.0  False   
2982     3225_01     Europa      True   C/120/S    55 Cancri e  17.0  False   
8586     9165_01      Earth     False  F/1771/S    TRAPPIST-1e  23.0  False   
7522     8043_03     Europa     False    A/97/S    TRAPPIST-1e  39.0   True   
1017     1085_01      Earth       NaN   G/165/P    TRAPPIST-1e  54.0    NaN   
4209     4483_02       Mars     False   D/151/P    TRAPPIST-1e  63.0  False   

     RoomService FoodCourt ShoppingMall    Spa  VRDeck                Name  \
3773         0.0       0.0          0.0    0.0     0.0   Terina Carpennels   
7018         0.0       0.0          0.0    0.0     0.0      Lynnie Hickett   
6898         1.0       0.0          0.0    0.0   718.0  Toniel Stroyledgez   
4157         0.0    4558.0          0.0    1.0   267.0    Weias Clinarfork   
1843         3.0       0.0         56.0  633.0     0.0     Megany Irwinson   
2982         0.0       0.0          0.0    0.0     0.0      Minoton Surind   
8586         9.0       2.0          1.0    0.0   727.0     Leria Cofferson   
7522         1.0     392.0         64.0  148.0  1367.0     Alcorix Fanseld   
1017         0.0       0.0          0.0    0.0     0.0    Troyce Wagnerray   
4209      3087.0       0.0         60.0    2.0     0.0        Fowoos Tepie   

     Transported  
3773       False  
7018        True  
6898        True  
4157        True  
1843       False  
2982        True  
8586       False  
7522        True  
1017       False  
4209       False

 Преобразования показателей. Обработайте каждый показатель в соответствии с тем типом данных, которым он должен быть (из первого пункта у нас должны быть изначально все показатели str типа)

In [7]:
df[['Age', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa', 'VRDeck', ]] = df[['Age', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa', 'VRDeck', ]].astype('float64')
# df['Transported'] = df['Transported'].astype('bool')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   object 
dtypes: float64(6), object(8)
memory usage: 950.9+ KB


In [8]:
df['Transported'].value_counts()

True     4378
False    4315
Name: Transported, dtype: int64

- Обработайте пропуски

In [9]:
df.shape

(8693, 14)

In [10]:
df['Age'] = df['Age'].fillna(df['Age'].mean())
df['HomePlanet'] = df['HomePlanet'].fillna(df['HomePlanet'].mode()[0])
df['RoomService'] = df['RoomService'].fillna(df['RoomService'].median())
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8693 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8693 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8693 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   object 
dtypes: float64(6), object(8)
memory usage: 950.9+ KB


Попытайтесь перейти от строкового типа данных к числовым

In [11]:
map_conditions = {
    'Europa': 1,
    'Earth': 2,
    'Mars': 3
}
df['HomePlanet'] = df['HomePlanet'].map(map_conditions)
df['HomePlanet'].unique()

array([1, 2, 3], dtype=int64)

- Посчитайте базовые статистики показателей используя группировку и встроенные методы

In [12]:
df.describe()

HomePlanet          Age   RoomService     FoodCourt  ShoppingMall  \
count  8693.000000  8693.000000   8693.000000   8510.000000   8485.000000   
mean      1.957207    28.827930    220.009318    458.077203    173.729169   
std       0.667613    14.339054    660.519050   1611.489240    604.696458   
min       1.000000     0.000000      0.000000      0.000000      0.000000   
25%       2.000000    20.000000      0.000000      0.000000      0.000000   
50%       2.000000    27.000000      0.000000      0.000000      0.000000   
75%       2.000000    37.000000     41.000000     76.000000     27.000000   
max       3.000000    79.000000  14327.000000  29813.000000  23492.000000   

                Spa        VRDeck  
count   8510.000000   8505.000000  
mean     311.138778    304.854791  
std     1136.705535   1145.717189  
min        0.000000      0.000000  
25%        0.000000      0.000000  
50%        0.000000      0.000000  
75%       59.000000     46.000000  
max    22408.000000  24133.000000

In [13]:
df[['HomePlanet', 'Age']].groupby('HomePlanet').agg(['mean','median'])

Age       
                 mean median
HomePlanet                  
1           34.298960   32.0
2           26.232878   24.0
3           29.285731   28.0

In [14]:
df[['Destination', 'Transported']].groupby(['Destination','Transported']).agg({'Transported': [ 'count']})

Transported
                                count
Destination   Transported            
55 Cancri e   False               702
              True               1098
PSO J318.5-22 False               395
              True                401
TRAPPIST-1e   False              3128
              True               2787

In [15]:
df['Transported'].value_counts(normalize=True)

True     0.503624
False    0.496376
Name: Transported, dtype: float64

In [16]:
df[['Destination']].value_counts(normalize=True)

Destination  
TRAPPIST-1e      0.694983
55 Cancri e      0.211491
PSO J318.5-22    0.093526
dtype: float64

In [17]:

df[['CryoSleep', 'Transported']].value_counts(normalize=True)

CryoSleep  Transported
False      False          0.430628
True       True           0.292945
False      True           0.211067
True       False          0.065361
dtype: float64

In [18]:
df['Transported'][df['CryoSleep'] == 'True'].value_counts(normalize=True)

True     0.817583
False    0.182417
Name: Transported, dtype: float64

In [23]:
df['Transported'][df['RoomService'] == 0].value_counts(normalize=True)

True     0.627822
False    0.372178
Name: Transported, dtype: float64

In [26]:
df['Transported'][df['RoomService'] > 0].value_counts(normalize=True)

False    0.740034
True     0.259966
Name: Transported, dtype: float64

In [30]:
df[['Transported', 'Age']].groupby('Transported').agg('describe').T

Transported        False         True
Age count    4315.000000  4378.000000
    mean       29.900275    27.771017
    std        13.558108    14.995903
    min         0.000000     0.000000
    25%        21.000000    18.000000
    50%        28.000000    27.000000
    75%        38.000000    37.000000
    max        79.000000    78.000000

Результаты изучения данных:
- более 50% пассажиров с планеты Earth;
- почти 70% пассажиров направлялись на TRAPPIST-1e;
- половина пассажиров пропали после столкновения;
- 80% пассажиров в криоспе переместились
- 62% тех кто не платил за комнаты пропали против 26% тех, кто платил, возможно это как то повлияло
- непохоже, что возраст как то сильно влиял на результат
